File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.py
from collections import deque
def solution(n, t, m, timetable):
    answer = 0
    start = 60*9
    end = 23*60 + 59
    
            
    def calc(a,b):
        return int(a)*60 + int(b)
    
    def re_calc(c):
        ah = str(c//60)
        am = str(c%60)
        if len(ah) == 1:
            ah = '0'+ah
        if len(am) == 1:
            am = '0'+ am
        tmp = ah+':'+am
        return tmp
    
    for i in range(len(timetable)):
        timetable[i] = calc(timetable[i][0:2], timetable[i][3:5])
        
    timetable.sort()
    q = deque(timetable)
    
    for _ in range(n):
        mx = m
        last = start
        while q:
            if mx!=0 and q[0] <= start:
                last = q.popleft()
                mx-=1
            else:
                break
        if _== n-1 and mx==0:
            answer = last -1
        elif _==n-1 and mx!=0:
            answer = start
                
        start += t       
        
    answer = re_calc(answer)

    return answer

File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.java
import java.util.*;
class Solution {
    public String solution(int n, int t, int m, String[] timetable) {
        String answer = "";
        
        List<Integer> times = new ArrayList<>();
        for(String tt: timetable){
            times.add(timeToMinute(tt));
        }
        Collections.sort(times);
        Queue<Integer> q = new LinkedList<>();
        int start = timeToMinute("09:00");
        int crew = 0;
        int last = 0;
        for(int i=0; i<n;i++){
            int cnt = 0;
            while(crew < times.size() && times.get(crew) <= start && cnt <m){
                last = times.get(crew);
                crew++;
                cnt++;
            }
            if(i==n-1){
                if(cnt<m){
                    answer = minuteToTime(start);
                }else{
                    answer = minuteToTime(last - 1);
                }
            }
            start+= t;
        }
        
        
        return answer;
    }
    public static int timeToMinute(String time){
        String[] tmp = time.split(":");
        return Integer.parseInt(tmp[0])*60+Integer.parseInt(tmp[1]);
    }
    public static String minuteToTime(int m) {
        int hour = m / 60;
        int min = m % 60;
    return String.format("%02d:%02d", hour, min);
}
}